obp/ope/estimators.py

```diff
@@ -537,6 +540,68 @@ def estimate_interval(
             random_state=random_state,
         )
 
+    def _estimate_mse_upper_bound(
+        self,
+        reward: np.ndarray,
+        action: np.ndarray,
+        pscore: np.ndarray,
+        action_dist: np.ndarray,
+        position: Optional[np.ndarray] = None,
+        max_reward_value: Optional[Union[int, float]] = None,
+    ) -> float:
+        """Estimate the upper bound of the mean-squared-error of IPW with a given clipping hyperparameter.
+
+        Parameters
+        ----------
+        reward: array-like, shape (n_rounds,)
+            Reward observed in each round of the logged bandit feedback, i.e., :math:`r_t`.
+
+        action: array-like, shape (n_rounds,)
+            Action sampled by a behavior policy in each round of the logged bandit feedback, i.e., :math:`a_t`.
+
+        pscore: array-like, shape (n_rounds,)
+            Action choice probabilities by a behavior policy (propensity scores), i.e., :math:`\\pi_b(a_t|x_t)`.
+
+        action_dist: array-like, shape (n_rounds, n_actions, len_list)
+            Action choice probabilities by the evaluation policy (can be deterministic), i.e., :math:`\\pi_e(a_t|x_t)`.
+
+        position: array-like, shape (n_rounds,), default=None
+            Positions of each round in the given logged bandit feedback.
+
+        Returns
+        ----------
+        estimated_mse_upper_bound: float
+            Estimated upper bound of MSE with a given clipping hyperparameter `lambda_`.
+            This is estimated using the automatic hyperparameter tuning procedure
+            based on Section 5 of Su et al.(2020).
+
+        """
+        n_rounds = reward.shape[0]
+        # estimate the variance of IPW with clipping
+        var_hat = np.var(
+            self._estimate_round_rewards(
+                reward=reward,
+                action=action,
+                pscore=pscore,
+                action_dist=action_dist,
+                position=position,
+            )
+        )
+        var_hat /= n_rounds
+
+        # estimate the upper bound of the bias of IPW with clipping
+        iw = action_dist[np.arange(n_rounds), action, position] / pscore
+        iw_hat = np.minimum(iw, self.lambda_)
+        bias_upper_bound_arr = (iw - iw_hat) * reward
+        bias_upper_bound = np.abs(bias_upper_bound_arr.mean())
+        bias_upper_bound += np.sqrt(
+            (2 * (iw ** 2).mean() * np.log(40)) / n_rounds
+        )  # \delta=0.05
+        bias_upper_bound += (2 * iw.max() * np.log(40)) / (3 * n_rounds)
```

Comment:
(Nitpick)
Maybe write a helper function that takes in `iw, iw_hat, reward, q_hat` as inputs and returns the bias upper bound? I think the formula is the same for all OPE estimators... (except for this one, for which you can pass in an array of 0s for `q_hat`)

Reply:
Fixed. Thanks!